### PR TITLE
Extend type coverage for relational and integer functions

### PR DESCRIPTION
**Pull request by kopylovanat**
Fixed ommiting tests for char and vec<char,N>

Remaining gap (test coverage): result correctness isn't checked

### CODE REVIEW

**Comment by bader** [comment]
@ProGTX, @illuhad, could you take a look, please?

**Review by bader — APPROVED** [review]

File: tests/math_builtin_api/modules/test_generator.py [hunk 1]
```diff
@@ -153,22 +154,26 @@ def generate_test_cases(test_id, types, sig_list):
 # We attempt to find a combination that is part of the current generic type (e.g. floatn)
 
 def attempt_match(runner, var_type, base_type, dim, unsigned, current_type):
+    ct_keys = list(current_type.keys())
+    ct_part_keys = []
+    for t in ct_keys:
```
Comment (keryell) [inline tests/math_builtin_api/modules/test_generator.py:159]: Would this work?
```suggestion
    for t in current_type.keys():
```

File: tests/math_builtin_api/modules/test_generator.py [hunk 2]
```diff
@@ -181,52 +186,71 @@ def expand_signature(runner, types, signature):
...
-                    if match is True:
-                        # Return value and all arguments are of the same type.
-                        if len(nomatch) == 0:
-                            common_type = current_types[0][(
-                                var_type, base_type, dim, unsigned)]
-                            new_sig = sycl_functions.funsig(signature.namespace, common_type.name, signature.name, [
-                                                            common_type.name for i in range(len(signature.arg_types))], signature.pntr_indx[:])
-                            exp_sig.append(new_sig)
+    all_types = sycl_types.create_basic_types()
+    for name in list(all_types.keys()):
```
Comment (keryell) [inline tests/math_builtin_api/modules/test_generator.py:192]:
```suggestion
    for name in all_types.keys():
```
?

File: tests/math_builtin_api/modules/test_generator.py [hunk 3]
```diff
@@ -181,52 +186,71 @@ def expand_signature(runner, types, signature):
...
+        match = False
+        nomatch = []
+        index = 0
+        for ct in current_types:
+            ct_keys = list(ct.keys())
```
Comment (keryell) [inline tests/math_builtin_api/modules/test_generator.py:197]: Needed?

File: tests/math_builtin_api/modules/test_generator.py [hunk 4]
```diff
@@ -181,52 +186,71 @@ def expand_signature(runner, types, signature):
...
+                new_sig = sycl_functions.funsig(signature.namespace, name, signature.name, [
+                                                name for i in range(len(signature.arg_types))], signature.pntr_indx[:])
+                exp_sig.append(new_sig)
+            else:
+                function_types = []
+                function_types_extra = []
+                all_matched = True
+                extra = False
+                for ct in current_types:
+                    ct_keys = list(ct.keys())
```
Comment (keryell) [inline tests/math_builtin_api/modules/test_generator.py:218]: Needed?

**Review by keryell — COMMENTED** [review]
Great!

File: tests/math_builtin_api/modules/test_generator.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 from . import sycl_functions
+from . import sycl_types
```
[the remark this answers is not in the document]
Comment (kopylovanat) [reply tests/math_builtin_api/modules/test_generator.py:2]: It's for getting all_types = sycl_types.create_basic_types()

File: tests/math_builtin_api/modules/test_generator.py (hunk 1 above)
Comment (kopylovanat) [reply tests/math_builtin_api/modules/test_generator.py]: Yes, it works, thank you! Removed list in other places too.

File: tests/math_builtin_api/modules/test_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (keryell) [reply tests/math_builtin_api/modules/test_generator.py:2]: Oh I see, there is a `__init__.py` in the directory.
I have never used `.` like that. https://docs.python.org/3/tutorial/modules.html#intra-package-references

**Review by keryell — APPROVED** [review]
Thanks.

**Comment by bader** [comment]
@AerialMantis, @ProGTX, @illuhad, I'd like to get at least one more approval from Codeplay team or Aksel to merge this PR.
The same applies to two more PRs: https://github.com/KhronosGroup/SYCL-CTS/pull/60 and https://github.com/KhronosGroup/SYCL-CTS/pull/61.

**Comment by bader** [comment]
> @AerialMantis, @ProGTX, @illuhad, I'd like to get at least one more approval from Codeplay team or Aksel to merge this PR.
> The same applies to two more PRs: #60 and #61.

Friendly ping.

File: tests/math_builtin_api/modules/test_generator.py [hunk 6]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
+            ct_keys = ct.keys()
+            ct_part_keys = []
+            for t in ct_keys:
+                ct_part_keys.append(t[4:])
+            if (name, ) in ct_part_keys:
+                match = True
+            else:
+                nomatch.append(index)
+            index += 1
+        if match is True:
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:205]:
```suggestion
        if match:
```

File: tests/math_builtin_api/modules/test_generator.py [hunk 7]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
+    for name in all_types.keys():
+        match = False
+        nomatch = []
+        index = 0
+        for ct in current_types:
+            ct_keys = ct.keys()
+            ct_part_keys = []
+            for t in ct_keys:
+                ct_part_keys.append(t[4:])
+            if (name, ) in ct_part_keys:
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:200]: I'm unsure about his, why not just:
```suggestion
            if name in ct_part_keys:
```

File: tests/math_builtin_api/modules/test_generator.py [hunk 8]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
+    all_types = sycl_types.create_basic_types()
+    for name in all_types.keys():
+        match = False
+        nomatch = []
+        index = 0
+        for ct in current_types:
+            ct_keys = ct.keys()
+            ct_part_keys = []
+            for t in ct_keys:
+                ct_part_keys.append(t[4:])
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:199]: Why `4`? Could this please have a comment, or maybe be stored in a named variable?

File: tests/math_builtin_api/modules/test_generator.py [hunk 9]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
+                        function_types_extra.append(name)
+                    else:
+                        var_type = all_types[name].var_type
+                        base_type = all_types[name].base_type
+                        dim = all_types[name].dim
+                        unsigned = all_types[name].unsigned
+
+                        # Get var_type, base_type, dim, unsigned of suitable type
+                        mutation = attempt_match(runner, var_type, base_type, dim, unsigned, ct)
+                        if mutation is not None:
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:233]:
```suggestion
                        if mutation:
```

File: tests/math_builtin_api/modules/test_generator.py [hunk 10]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
-                                        all_matched = False
-                            if all_matched is True:
-                                new_sig = sycl_functions.funsig(
-                                    signature.namespace, function_types[0], signature.name, function_types[1:], signature.pntr_indx[:])
-                                exp_sig.append(new_sig)
-                            else:
-                                print("[WARNING] Unable to fully match function " + signature.name + " for: " +
-                                      var_type + ", " + base_type + ", " + str(dim) + ", " + str(unsigned))
+                            all_matched = False
+                if all_matched is True:
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:242]:
```suggestion
                if all_matched:
```

File: tests/math_builtin_api/modules/test_generator.py [hunk 11]
```diff
@@ -181,52 +185,71 @@ def expand_signature(runner, types, signature):
...
+            else:
+                function_types = []
+                function_types_extra = []
+                all_matched = True
+                extra = False
+                for ct in current_types:
+                    ct_keys = ct.keys()
+                    ct_part_keys = []
+                    for t in ct_keys:
+                        ct_part_keys.append(t[4:])
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:220]: There's another magic number here, I think a common variable is the best bet then

File: tests/math_builtin_api/modules/test_generator.py [hunk 12]
```diff
@@ -153,22 +154,25 @@ def generate_test_cases(test_id, types, sig_list):
 # We attempt to find a combination that is part of the current generic type (e.g. floatn)
 
 def attempt_match(runner, var_type, base_type, dim, unsigned, current_type):
+    ct_part_keys = []
+    for t in current_type.keys():
+        ct_part_keys.append(t[:-1])
```
Comment (ProGTX) [inline tests/math_builtin_api/modules/test_generator.py:159]: Can a comment please be added indicating why we skip the last value?

**Review by ProGTX — CHANGES_REQUESTED** [review]
Sorry for not reviewing this earlier. The changes mostly seem like a welcome addition, I've left some suggestions.

What's more worrying is that ComputeCpp fails to compile the newly generated file, specifically this is where the first error happens:
```cpp
cl::sycl::char2 inputData_0(58,-109);
signed char inputData_1(90);
return cl::sycl::max(inputData_0,inputData_1);
```

I'm not sure if the SYCL spec supports this use case, or whether it should. `std::max` will also fail to compile if you pass in a `char` and a `signed char`. Even if that were defined, I think the result might depend on whether `char` itself is signed or not. Does this work in DPC++, though?

I don't know exactly which builtins are impacted by this, but it seems like there might need to be more restrictions when generating the file.

**Comment by bader** [comment]
>Even if that were defined, I think the result might depend on whether `char` itself is signed or not.

Unless I missed something, it seems to me that the spec doesn't allow ambiguity for this case. According to the SYCL-1.2.1 specification: 

> `cl_char` - Alias to a signed 8-bit integer, as defined bythe C++11 ISO Standard

`cl::sycl::cl_char2` is expected to be an alias to 2-element vector of `signed int8_t` or `signed char` (if `char` == `int8_t`).

**Comment by keryell** [comment]
> > Even if that were defined, I think the result might depend on whether `char` itself is signed or not.
> 
> Unless I missed something, it seems to me that the spec doesn't allow ambiguity for this case. According to the SYCL-1.2.1 specification:
> 
> > `cl_char` - Alias to a signed 8-bit integer, as defined bythe C++11 ISO Standard
> 
> `cl::sycl::cl_char2` is expected to be an alias to 2-element vector of `int8_t` or `signed char` (if `sizeof(char)` == `sizeof(int8_t)`).

I think you are right but then why not using `std::int8_t` to be clearer and to fail in the rare case a `signed char` has not 8 bits?

**Comment by ProGTX** [comment]
I combed through the 1.2.1 and 2020 provisional specs and discussed this internally a bit.

This is what the spec (both of them) says:
```cpp
geninteger max (geninteger x, geninteger y)
geninteger max (geninteger x, sgeninteger y)
```
> The built-in functions can take as input char, unsigned char, short, unsigned short, int, unsigned int, long long int,
unsigned long long int and their vec and marray counterparts.

This is the conclusion we came to:
1. SYCL builtins are defined in terms of C++ types, not OpenCL ones
2. `max(scalar, vector)` is probably fine, though that might need a clarification in the spec
3. `max(signed, unsigned)` is fine
4. `max(char, (un)signed char)` is not allowed because the result would not be portable
5. The spec doesn't even mention `signed char` for the builtins

So in theory this patch wouldn't be correct because of point 5 alone. But we believe that to be an oversight, so we might want to fix the spec.

Point 4 is the problematic one, though. Specifically the example I posted already gives different results on X86 and ARM, for example.

**Comment by bader** [comment]
> 3. `max(signed, unsigned)` **is fine**
> 5. The spec **doesn't even mention** `signed char` for the builtins

Are there any conflicts between these items?

**Comment by ProGTX** [comment]
I meant `max(signed, unsigned)` for the generic case, so `max(int, unsigned int)` should be fine, and I believe `max(signed char, unsigned char)` would be as well if the spec mentioned `signed char`. `char` is a distint type from both, though, so I don't think there's a conflict there.

**Comment by bader** [comment]
> I meant `max(signed, unsigned)` for the generic case, so `max(int, unsigned int)` should be fine, and I believe `max(signed char, unsigned char)` would be as well if the spec mentioned `signed char`. `char` is a distint type from both, though, so I don't think there's a conflict there.

I agree. Thanks for clarification.

**Comment by kopylovanat** [comment]
@ProGTX thank you for comments! Regarding `max` with `char` issue:

> 2. `max(scalar, vector)` is probably fine, though that might need a clarification in the spec

`max(vector, scalar)` is fine, because it corresponds to `max (geninteger x, sgeninteger y)`
> 3. `max(signed, unsigned)` is fine

From my understanding it's not, because in `max (geninteger x, geninteger y)` x and y types should be the same and in `max (geninteger x, sgeninteger y)` x and y types should be vector and scalar with the same base type.
> 4. `max(char, (un)signed char)` is not allowed because the result would not be portable

And this case also shouldn't be allowed with the same reasoning as for 3.
> 5. The spec doesn't even mention `signed char` for the builtins
> 
> So in theory this patch wouldn't be correct because of point 5 alone. But we believe that to be an oversight, so we might want to fix the spec.

It's too seems to me as spec for integer functions missed `signed char` because generic type `geninteger` contains `signed char` according to Table 4.109

> 
> Point 4 is the problematic one, though. Specifically the example I posted already gives different results on X86 and ARM, for example.

Example failed because it has `vec<char,2>` and `signed char` as arguments, so it's vector and scalar with different base types. And it's shouldn't be tested (even though it works on DPC++ because `char` and `signed char` are treated as the same). I'll modify generation code to stop these cases from being generated.

File: tests/math_builtin_api/modules/test_generator.py (hunk 9 above)
Comment (kopylovanat) [reply tests/math_builtin_api/modules/test_generator.py]: Done.

File: tests/math_builtin_api/modules/test_generator.py (hunk 10 above)
Comment (kopylovanat) [reply tests/math_builtin_api/modules/test_generator.py]: Done.

**Comment by bader** [comment]
@ProGTX, could you check if https://github.com/KhronosGroup/SYCL-CTS/pull/59/commits/95d4c475decae999cc49cb38de30354f069ea099 addresses your comments, please? Let me know if any other changes or other actions are required.

**Review by ProGTX — APPROVED** [review]
You're right about `max(signed, unsigned)` not being valid and the implications of that.

The code works now with ComputeCpp, thanks for the changes.

**Pull request merged**